### PR TITLE
[SPARK-19952][SQL] Remove various analysis exceptions

### PR DESCRIPTION
**Pull request by hvanhovell**
## What changes were proposed in this pull request?
We currently have quite a few analysis exception subclasses, the problem with these is that they are not well supported throughout Spark (in pySpark for example) and that the added value is limited. This PR removes the `no such object` and the `already exists` exceptions, and replaces them by a factory methods that create analysis exceptions with the required error messages.

## How was this patch tested?
Modified existing tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -636,17 +636,16 @@ class Analyzer(
         defaultDatabase: Option[String] = None): LogicalPlan = {
       val tableIdentWithDb = u.tableIdentifier.copy(
         database = u.tableIdentifier.database.orElse(defaultDatabase))
-      try {
-        catalog.lookupRelation(tableIdentWithDb)
-      } catch {
-        case _: NoSuchTableException =>
-          u.failAnalysis(s"Table or view not found: ${tableIdentWithDb.unquotedString}")
-        // If the database is defined and that database is not found, throw an AnalysisException.
+      if (!tableIdentWithDb.database.exists(catalog.databaseExists)) {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Perhaps we should move this into `catalog.lookupRelation` and attach the location here.

**Comment by hvanhovell** [comment]
cc @cloud-fan @ueshin @rxin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: how about
```
try {
  catalog.lookupRelation(tableIdentWithDb)
} catch {
  case a: AnalysisException => u.failAnalysis(a.message)
}
```

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yeah, could work. This change is also causing the tests to fail, so I am revisiting it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 2]
```diff
@@ -160,7 +159,7 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
   }
 
   private def assertNoSuchTable(query: String): Unit = {
-    intercept[NoSuchTableException] {
+    intercept[AnalysisException] {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: `AnalysisException` is a pretty general Exception in Spark SQL. The future refactoring might introduce bugs without trigger any test case failure. 

Could we check the error messages?

Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: Will do

**Comment by HyukjinKwon** [comment]
@hvanhovell, how is it going?

**Comment by jiangxb1987** [comment]
ping @hvanhovell Are you still working on this?

**Pull request closed without merging**